build.assets/build-webassets-if-changed.sh

```diff
@@ -0,0 +1,84 @@
+#!/bin/bash
+#
+# Rebuilds a UI make target based on whether the given directories contents have changed
+# minus any node-modules.
+#
+
+set -eo pipefail
+
+ROOT_PATH="$(realpath "$(dirname "$0")/..")"
+MAKE="${MAKE:-make}"
+SHASUMS=("shasum -a 512" "sha512sum" "sha256sum")
+
+if ! command -v "$MAKE" >/dev/null; then
+  echo "Unable to find \"$MAKE\" on path."
+  exit 1
+fi
+
+if [ -n "$SHASUM" ]; then
+  EXEC="$(echo "$SHASUM" | awk '{print $1}')"
+  if ! command -v "$EXEC" >/dev/null; then
+    echo "Unable to find custom SHA sum $SHASUM on path."
+    exit 1
+  fi
+else
+  for shasum in "${SHASUMS[@]}"; do
+    EXEC="$(echo "$shasum" | awk '{print $1}')"
+    if command -v "$EXEC" >/dev/null; then
+      SHASUM="$shasum"
+      break
+    fi
+  done
+fi
+
+if [ -z "$SHASUM" ]; then
+  echo "Unable to find a SHA sum executable."
+  exit 1
+fi
+
+if [ "$#" -lt 4 ]; then
+  echo "Usage: $0 <type> <last-sha-file> <build-target> <directories...>"
+  exit 1
+fi
+
+TYPE="$1"
+LAST_SHA_FILE="$ROOT_PATH/$2"
+BUILD_TARGET="$3"
+shift 3
+SRC_DIRECTORIES=("$@")
+
+for i in "${!SRC_DIRECTORIES[@]}"; do
+  SRC_DIRECTORIES[i]="$ROOT_PATH/${SRC_DIRECTORIES[i]}"
+done
+
+function calculate_sha() {
+  #shellcheck disable=SC2086
+  find "${SRC_DIRECTORIES[@]}" "$ROOT_PATH/package.json" "$ROOT_PATH/yarn.lock" | grep -v 'node_modules' | LC_ALL=C sort | cpio -o 2>/dev/null | $SHASUM | cut -f1 -d' '
```

Comment:
Avoid searching through `node_modules`:
```suggestion
  find "${SRC_DIRECTORIES[@]}" "$ROOT_PATH/package.json" "$ROOT_PATH/yarn.lock" -not \( -type d -name node_modules -prune \) | LC_ALL=C sort | cpio -o 2>/dev/null | $SHASUM | cut -f1 -d' '
```

Use NUL to separate filenames, to handle files with spaces in them correctly:
```suggestion
  find "${SRC_DIRECTORIES[@]}" "$ROOT_PATH/package.json" "$ROOT_PATH/yarn.lock" -not \( -type d -name node_modules -prune \) -print0 | LC_ALL=C sort -z | cpio -0 -o 2>/dev/null | $SHASUM | cut -f1 -d' '
```

Reply:
That being said, `cpio` will record mtime (and ctime?) for each file, so even touching a file will result in a different hash.

Reply:
I think avoiding that requires too much in way of controversial build dependencies, so I think this is a reasonable approach. I've taken the `print0` modifications.